### PR TITLE
Use `wasmImports` as the single global symbol table for dynamic linking

### PR DESCRIPTION
**Pull request by sbc100**
Previously we were looking up symbols on the `Module` object which meant building with `EXPORT_ALL` and exporting all symbols there.  This change adds all needed symbols to the `wasmImports` map and uses that as the single source for symbol lookups.

This is split out from my work on #18376.

This should be a code size win for most users of dynamic linking.  It also avoids extra symbols being present on the `Module` object that the user didn't ask for.

### CODE REVIEW

File: emscripten.py [hunk 1]
```diff
@@ -685,6 +684,30 @@ def create_sending(invoke_funcs, metadata):
...
+      for f in library_symbols:
+        if shared.is_c_symbol(f):
+          demangled = shared.demangle_c_symbol_name(f)
+          if demangled in exports:
+            continue
+          send_items_map[demangled] = f
+    else:
+      for f in settings.EXPORTED_FUNCTIONS + settings.SIDE_MODULE_IMPORTS:
+        if f in library_symbols and shared.is_c_symbol(f):
+          if f == '_main' and '__main_argc_argv' in exports:
```
Comment (kripken) [inline emscripten.py:704]: Maybe add a comment about why main is special? That's not obvious to me.

File: src/library_dylink.js [hunk 2]
```diff
@@ -463,38 +454,43 @@ var LibraryDylink = {
...
+        continue;
       }
 #endif
 
-      // Export native export on the Module object.
-      // TODO(sbc): Do all users want this?  Should we skip this by default?
-      var module_sym = asmjsMangle(sym);
-      if (!Module.hasOwnProperty(module_sym)) {
-        Module[module_sym] = exports[sym];
+      // When RTLD_GLOBAL is enable, the symbols defined by this shared object will be made
```
Comment (kripken) [inline src/library_dylink.js:468]:
```suggestion
      // When RTLD_GLOBAL is enabled, the symbols defined by this shared object will be made
```

File: src/jsifier.js [hunk 3]
```diff
@@ -426,6 +426,7 @@ function ${name}(${args}) {
       }
       if (isStub) {
         contentText += `\n${mangled}.stub = true;`;
+        contentText += `\nwasmImportStubs['${symbol}'] = undefined;`;
```
Comment (kripken) [inline src/jsifier.js:429]: Why do we need to set it to undefined?

File: src/library_exceptions.js [hunk 4]
```diff
@@ -121,7 +121,7 @@ var LibraryExceptions = {
     this.get_exception_ptr = function() {
       // Work around a fastcomp bug, this code is still included for some reason in a build without
       // exceptions support.
-      var isPointer = {{{ exportedAsmFunc('___cxa_is_pointer_type') }}}(this.get_type());
```
Comment (kripken) [inline src/library_exceptions.js:124]: Is this change related?

Comment (sbc100) [reply src/library_exceptions.js:124]: Its related, since now we simply don't need `exportedAsmFunc`.     However, I wonder if I can split this part out..

Comment (sbc100) [reply src/library_exceptions.js:124]: Spit out into #18588

Comment (sbc100) [reply src/library_exceptions.js]: Rebased.  This is no longer part of this change.

File: src/jsifier.js (hunk 3 above)
Comment (sbc100) [reply src/jsifier.js:429]: So that `library_dylink.js` knows to full this in with the real symbol later.

I will add some comments about this.

File: emscripten.py (hunk 1 above)
Comment (sbc100) [reply emscripten.py]: I removed these lines!

File: src/preamble.js [hunk 5]
```diff
@@ -1223,4 +1223,13 @@ function runMemoryInitializer() {
 }
 #endif // MEM_INIT_IN_WASM == 0
 
+#if MAIN_MODULE && ASYNCIFY
+// With MAIN_MODULE + ASYNCIFY the normal method of placing stub functions in
+// wasmImports for as-yet-undefined symbols doesn't work since ASYNCIFY then
+// wraps these stub functions and we can't then replace them directly.  Instead
+// the stub functions call into `asyncifyStubs` which gets populated by the
+// dyanmic linker and symbols are loaded.
```
Comment (kripken) [inline src/preamble.js:1231]:
```suggestion
// dynamic linker and symbols are loaded.
```

Comment (kripken) [reply src/preamble.js:1231]: Should "and" be "when" perhaps?

File: src/jsifier.js [hunk 6]
```diff
@@ -271,10 +271,13 @@ function ${name}(${args}) {
           // (not useful to warn/error multiple times)
           LibraryManager.library[symbol + '__docs'] = '/** @type {function(...*):?} */';
         } else {
-          const target = `Module['${mangled}']`;
+          let target = `wasmImports['${symbol}']`;
+          if (ASYNCIFY) {
+            target = `asyncifyStubs['${symbol}']`;
```
Comment (kripken) [inline src/jsifier.js:276]: Other places in the code seem to suggest that a symbol might or might not be in `asyncifyStubs`, but here we read from it unconditionally and overwrite the old value. Is that intentional?

Comment (sbc100) [reply src/jsifier.js:276]: Here we generating the stub for missing `symbol`.   And we using `asyncifyStubs['${symbol}']` only within that stub function itself.

If this stub function is called, and `asyncifyStubs['${symbol}']` is still undefined at that point in time, that would be critical error and the program will abort.   In debug builds we show the assert below in that case.

Comment (kripken) [inline src/jsifier.js:276]:
```suggestion
          let target;
          if (!ASYNCIFY) {
            target = `wasmImports['${symbol}']`;
          } else {
            target = `asyncifyStubs['${symbol}']`;
```

Comment (kripken) [reply src/jsifier.js:276]: I see, sounds good. In that case I wrote a suggestion to revise the code slightly.

File: src/library_dylink.js [hunk 7]
```diff
@@ -10,40 +10,40 @@ var dlopenMissingError = "'To use dlopen, you need enable dynamic linking, see h
 
 var LibraryDylink = {
 #if RELOCATABLE
+  $isSymbolDefined: function(symName) {
+    // Ignore 'stub' symbols that are auto-generated as part of the original
+    // `wasmImports` used to instantate the main module.
+    var existing = wasmImports[symName];
+    if (!existing || existing.stub) return false;
```
Comment (kripken) [inline src/library_dylink.js:17]:
```suggestion
    if (!existing || existing.stub) {
      return false;
    }
```

File: src/library_dylink.js [hunk 8]
```diff
@@ -10,40 +10,40 @@ var dlopenMissingError = "'To use dlopen, you need enable dynamic linking, see h
...
+#if ASYNCIFY
+    // Even if a symbol exists in wasmImports and it not a stub, it could
+    // be an ASYNCIFY wrapper function.
+    if (symName in asyncifyStubs && !asyncifyStubs[symName]) return false;
```
Comment (kripken) [inline src/library_dylink.js:21]:
```suggestion
    if (symName in asyncifyStubs && !asyncifyStubs[symName]) {
      return false;
    }
```

File: src/library_dylink.js [hunk 9]
```diff
@@ -10,40 +10,40 @@ var dlopenMissingError = "'To use dlopen, you need enable dynamic linking, see h
 
 var LibraryDylink = {
 #if RELOCATABLE
+  $isSymbolDefined: function(symName) {
+    // Ignore 'stub' symbols that are auto-generated as part of the original
+    // `wasmImports` used to instantate the main module.
+    var existing = wasmImports[symName];
+    if (!existing || existing.stub) return false;
+#if ASYNCIFY
+    // Even if a symbol exists in wasmImports and it not a stub, it could
```
Comment (kripken) [inline src/library_dylink.js:19]:
```suggestion
    // Even if a symbol exists in wasmImports and it is not a stub, it could
```

File: src/preamble.js (hunk 5 above)
Comment (kripken) [inline src/preamble.js:1231]:
```suggestion
// dynamic linker as symbols are loaded.
```
is this right?

Comment (kripken) [inline src/preamble.js:1231]: Which functions end up here? I thought all of them, initially, but reading other code I am not sure.

Comment (kripken) [reply src/preamble.js:1231]: That is, one place seems to always read from `asyncifyStubs`, but another place checks if a thing is there or not, so I am confused.

Comment (sbc100) [reply src/preamble.js:1231]: Only functions for which we create stubs need to appear in this map.   Its the same set of functions that get the `.stub` attribute set up them.  These are functions which are required to instantiate the main module, but which are provided later, by a side module.

Without ASYNCIFY we simply replace the existing `.stub` function wasmImports once the symbols gets resolved.
With ASYNCIFY this appraoch doesn't work because the thing in the wasmImports dict might be an asyncify wrapper around the `.stub` we created.   In this case we replace the entry in `asyncifyStubs` which is called by the stub function.

Comment (sbc100) [reply src/preamble.js]: There are two places the `asyncifyStubs` are used.

1. In the stub functions themselves.   In debug builds we assert that that right entry is present.  In release builds we will just trap if its not present.
2. In the dynamic linker: When a new symbol is loaded it replaces the entry in asyncifyStubs, but only if that key already exists in the dictionary.

File: tools/building.py [hunk 10]
```diff
@@ -823,7 +823,10 @@ def metadce(js_file, wasm_file, minify_whitespace, debug_info):
   import_name_map = {}
   for item in graph:
     if 'import' in item:
-      import_name_map[item['name']] = 'emcc$import$' + item['import'][1]
+      name = item['import'][1]
+      import_name_map[item['name']] = 'emcc$import$' + name
+      if asmjs_mangle(name) in settings.SIDE_MODULE_IMPORTS:
```
Comment (kripken) [inline tools/building.py:828]: Why do we root side module imports? If we don't use such an import can't we drop it?

**Review by kripken — APPROVED** [review]
Thanks, that's what I was missing. Code makes sense to me now...

File: tools/building.py (hunk 10 above)
Comment (sbc100) [reply tools/building.py:828]: The side module requires them .. so we can't delete them.  Think JS library functions that are only used by the side module.  

For example, imagine that both the main module and the side module call `my_js_func`.   without this code, DCE, might notice that the main module usage can be removed and therefore try to also completely from `my_js_func`.

Historically this has always been the case since we were running with EXPORT_ALL which prevents any DCE of JS functions since they are all exported on the Module.

Comment (sbc100) [reply tools/building.py:828]: In case this isn't clear `SIDE_MODULE_IMPORTS` means "symbols imported/required by the side module(s)"

Comment (kripken) [reply tools/building.py:828]: Oh right, sorry... I read the docs for `SIDE_MODULE_EXPORTS` and not `_IMPORTS`...
